File: webui/src/models/Article.ts
import ArticleTag from "./ArticleTag"

export const Kind = {
  Markdown: 'markdown',
  Tweet: 'tweet',
  SlideShare: 'slideshare',
  Youtube: 'youtube',
}

export default class Article {
  id: number
  kind: string
  url: string
  _title: string
  tags: ArticleTag[]
  created: Date
  lastModified: Date

  constructor(obj: any) {
    this.id = obj.id
    this.kind = obj.kind
    this.url = obj.url
    this._title = obj.title
    this.tags = obj.tags
    this.created = new Date(obj.created)
    this.lastModified = new Date(obj.lastModified)
  }

  get title(): string {
    return this._title || '[empty title]'
  }

  set title(value: string) {
    this._title = value
  }
}

File: models/article.go
package models

import (
	"github.com/jaeyo/personal-archive/common"
	"gorm.io/gorm"
	"time"
)

const (
	KindMarkdown   = "markdown"
	KindTweet      = "tweet"
	KindSlideShare = "slideshare"
	KindYoutube    = "youtube"
)

type Article struct {
	ID           int64       `gorm:"column:id;primarykey" json:"id"`
	Kind         string      `gorm:"column:kind;type:varchar(24);not null" json:"kind"`
	URL          string      `gorm:"column:url;type:varchar(256);not null" json:"url"`
	Content      string      `gorm:"column:content;type:text" json:"content"`
	Title        string      `gorm:"column:title;type:varchar(256);not null;uniqueIndex" json:"title"`
	Tags         ArticleTags `gorm:"foreignKey:ArticleID" json:"tags"`
	Created      time.Time   `gorm:"column:created;type:datetime;not null" json:"created"`
	LastModified time.Time   `gorm:"column:last_modified;type:datetime;not null" json:"lastModified"`
}

func NewArticle(kind, url, content, title string, tags []string) *Article {
	articleTags := common.Map(tags, func(item string) *ArticleTag { return &ArticleTag{Tag: item} })

	return &Article{
		Kind:    kind,
		URL:     url,
		Content: content,
		Title:   title,
		Tags:    articleTags,
	}
}

func (a *Article) TableName() string {
	return "article"
}

func (a *Article) BeforeSave(db *gorm.DB) error {
	if a.Created.IsZero() {
		a.Created = time.Now()
	}
	a.LastModified = time.Now()
	return nil
}
